controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java

```diff
@@ -311,12 +312,16 @@ public static String credentialReferencePartAsStringIfDefined(ModelNode credenti
             secret = null;
         }
 
-        final InjectedValue<CredentialStore> credentialStoreInjectedValue = new InjectedValue<>();
+        final Service<CredentialStore> credentialStoreService;
         if (credentialAlias != null) {
             // use credential store service
             String credentialStoreCapabilityName = RuntimeCapability.buildDynamicCapabilityName(CREDENTIAL_STORE_CAPABILITY, credentialStoreName);
             ServiceName credentialStoreServiceName = context.getCapabilityServiceName(credentialStoreCapabilityName, CredentialStore.class);
-            serviceBuilder.addDependency(ServiceBuilder.DependencyType.REQUIRED, credentialStoreServiceName, CredentialStore.class, credentialStoreInjectedValue);
+            serviceBuilder.addDependency(ServiceBuilder.DependencyType.REQUIRED, credentialStoreServiceName);
+            ServiceController<?> controller = context.getServiceRegistry(false).getService(credentialStoreServiceName);
+            credentialStoreService = controller != null ? (Service<CredentialStore>)controller.getService() : null;
```

Comment:
This is a race. There is no guarantee that when L321 executes the credential store service will be installed in MSC. Subsystems execute their operations in parallel during boot so subsystem A can in no way rely on subystem B having executed its ops first. And even without parallel boot there its totally valid for users to place the elytron subsystem last in their config.